PTL-5574 Adds the ability to navigate emoji picker with keyboard

stephenkiers opened the pull request:
#### Summary
Adds the ability to navigate emoji picker with keyboard

#### Ticket Link
https://mattermost.atlassian.net/browse/PLT-5574
https://mattermost.atlassian.net/browse/PLT-8482

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] Ran `make check-style` to check for style errors (required for all pull requests)
- [x] Ran `make test` to ensure unit and component tests passed
- [x] Has UI changes

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

stephenkiers commented:
This is a re-merge of the emoji picker. I also fixed two issues that had come up that prevented it from being included in 4.6.

1- There was a component key collision issue that was happening causing weird search results and ux issues. I fixed this by making the key more specific when it is a custom emoji.
Issue probably would never have happened in the wild, but it is now fixed. 
re: https://mattermost.atlassian.net/browse/PLT-8482
2- There was an issue using scrollIntoView when you were adding a reaction. This has now been fixed.

Let me know if there is anything else required. 👍

mattermod commented:
Spinmint test server created at: http://i-00926a1cc1b480bcc.spinmint.com

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-00926a1cc1b480bcc

stephenkiers commented:
I think that is a separate issue and pull request https://mattermost.atlassian.net/browse/PLT-8384

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-09c4bfb68e11ea0e6.spinmint.com

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-09c4bfb68e11ea0e6

saturninoabril commented on components/emoji/components/emoji_list.jsx:
```diff
@@ -16,6 +16,10 @@ import LoadingScreen from 'components/loading_screen.jsx';
 
 import EmojiListItem from './emoji_list_item.jsx';
 
+function deleteEmoji(emoji) {
```
Why not inside the class?

stephenkiers replied:
My reasoning was because it is a static method that gets re-instantiated for every instance of the class. Outside the class it gets instantiated at runtime, and reused for every class. Perhaps it is a premature optimization though and shouldn't have been done though without testing.
It is more of a style issue than a major change.

stephenkiers replied:
Do you want this reverted, or is it good this way? Waiting on this to be resolved to mark ready for merge

saturninoabril replied:
If this is more of style/pattern, then I would prefer to have it inside the class like what we're doing elsewhere.

stephenkiers replied:
I will change this back, because I am only 1/5 on it.

 However, I am curious how decisions like this are made? Is it simply "how we always did it", or is there a proper style guide or decision making framework to guide these calls? I simply ask because I would hate for us to make poor decisions habitually just because they are the same as other outdated or poor decisions.

saturninoabril replied:
I think you may open up your idea or opinion in any of the discussion channel like developer channel, or might good to open a topic in developer meeting so you will hear instant feedback/discussion/agreement among developers.

stephenkiers replied:
great idea. I will look into that.

saturninoabril commented on components/emoji_picker/components/emoji_picker_item.jsx:
```diff
@@ -3,77 +3,95 @@
 
 import EmojiStore from 'stores/emoji_store.jsx';
 
+const SCROLLING_ADDT_VISUAL_SPACING = 10; // to make give the emoji some visual 'breathing room'
```
What does ADDT means? Addition?

stephenkiers replied:
additional

stephenkiers commented:
Regarding lazy loading, this does do that. https://www.youtube.com/watch?v=QF82L_jUlxI

saturninoabril commented:
There seems to be out of sync between emoji category and emoji item after doing a search. Not sure what's the expected behavior but I think highlighting of category on search is disabled on scroll since emoji results are not group according to category.

![jan-16-2018 01-49-04](https://user-images.githubusercontent.com/5334504/34955477-83883ab2-fa5f-11e7-96b0-018993ea528a.gif)

stephenkiers commented:
Good catch on the search and category highlighting issue. I will fix that quick

mattermod commented:
Spinmint test server destroyed

stephenkiers commented:
@saturninoabril What did you mean by "Loading on hover (previously applied by CB) could easily reapply again." in chat. Is there a feature I regressed unintentionally?

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-06dd3b46f1faf4e9c.spinmint.com

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-06dd3b46f1faf4e9c

saturninoabril commented:
No, it's not here and not a regression either. It was maybe on version below 4.2 which was removed in favor of current lazy loading.
Such feature was loading of emojis of emoji picker on hover of emoji icon (at create post, create comment, dot menu).
I was just thinking that your implementation of lazy loading is a good fit to bring it back. (Worth separate PR as another improvement).

stephenkiers commented:
yeah. I think there are some major performance updates that could be made to this editor. However, that will be another PR. I think the Performance team are going to be looking at improving it soon as well. One of our clients has over 6000 emojis and that really slows it down. (as you know. ;) )

stephenkiers commented:
@saturninoabril Since this isn't a part of a release, am I able to merge, or do we just add the tag? #protocolQuestion

saturninoabril commented:
@stephenkiers Seems late for me to reply but it's always require to have 2 dev approvals before doing a merge
- list of [core developers](https://docs.mattermost.com/developer/core-developer-handbook.html#current-core-developers)
- [managing an open PR](https://docs.mattermost.com/developer/contribution-guide.html#managing-an-open-pull-request).
